lib/block_view_message.go

```diff
@@ -218,19 +396,124 @@ func (bav *UtxoView) _connectPrivateMessage(
 	// At this point the inputs and outputs have been processed. Now we
 	// need to handle the metadata.
 
+	//Check if message is encrypted with shared secret
+	version := uint64(1)
+	if extraV, hasExtraV := txn.ExtraData["V"]; hasExtraV {
+		rr := bytes.NewReader(extraV)
+		version, err = ReadUvarint(rr)
+		if err != nil {
+			return 0, 0, nil, errors.Wrapf(RuleErrorPrivateMessageInvalidVersion,
+				"_connectPrivateMessage: Problem reading version from ExtraData, error: (%v)", err)
+		}
+		if version > 3 {
+			return 0, 0, nil, errors.Wrapf(RuleErrorPrivateMessageInvalidVersion,
+				"_connectPrivateMessage: Problem reading version from ExtraData")
+		}
+	}
+
+	// Create a MessageEntry
+	messageEntry := &MessageEntry{
+		SenderPublicKey:             NewPublicKey(txn.PublicKey),
+		RecipientPublicKey:          NewPublicKey(txMeta.RecipientPublicKey),
+		EncryptedText:               txMeta.EncryptedText,
+		TstampNanos:                 txMeta.TimestampNanos,
+		Version:                     uint8(version),
+		SenderMessagingPublicKey:    NewPublicKey(txn.PublicKey),
+		SenderMessagingKeyName:      BaseKeyName(),
+		RecipientMessagingPublicKey: NewPublicKey(txMeta.RecipientPublicKey),
+		RecipientMessagingKeyName:   BaseKeyName(),
+	}
+
+	// If message was encrypted using DeSo V3 Messages, we will look for messaging keys in
+	// ExtraData. V3 allows users to register messaging keys on-chain, and encrypt messages
+	// to these messaging keys, as opposed to encrypting to user's main keys.
+	if version == 3 {
+		// Make sure DeSo V3 messages are live.
+		if blockHeight < DeSoV3MessagesBlockHeight {
+			return 0, 0, nil, errors.Wrapf(
+				RuleErrorPrivateMessageMessagingPartyBeforeBlockHeight,
+				"_connectPrivateMessage: messaging party used before block height")
+		}
+		// Look for messaging keys in transaction ExtraData
+		// TODO: Do we want to make the ExtraData keys shorter to save space and transaction cost?
```

Comment:
Probably a good idea to keep the key length short if we can help it while maintaining some level of significance in the key name

Reply:
Yeaaah, I wonder what we could rename them to. @diamondhands0 what do you think about this one?